src/backend/executor/execTuples.c

```diff
@@ -751,6 +751,53 @@ ExecFetchSlotHeapTuple(TupleTableSlot *slot)
 	return htup;
 }
 
+HeapTuple
+ExecMaterializeSlotHeapTuple(TupleTableSlot *slot)
+{
+	uint32 tuplen;
+	HeapTuple htup;
+
+	/*
+	 * sanity checks
+	 */
+	Assert(!TupIsNull(slot));
+
+	/*
+	 * If we have a regular physical tuple, and it's locally palloc'd, we have
+	 * nothing to do.
+	 */
+	if (slot->PRIVATE_tts_heaptuple && slot->PRIVATE_tts_htup_buf)
```

Comment:
Should we use TupShouldFree logic here?

Reply:
I'm not convinced that `TupShouldFree` does the same thing here as the `shouldFree` flag in upstream. For instance, "materialized" slots that go through `ExecFetchSlotHeapTuple` don't seem to have their `TTS_SHOULDFREE` flag set. But I agree that my solution does not seem to be the best way to check for it...

Reply:
Right. To me, that seems like a reason *not* to use TupShouldFree, because `SHOULD_FREE` doesn't mean the same thing as "locally palloc'd". Am I missing anything?